File: MicroCli/src/main/resources/micronaut/security/java/mongoReactive/services/CustomRefreshTokenPersistence.java
package ${securityPackage}.services;


import ${securityPackage}.domains.RefreshToken;
import ${securityPackage}.repository.RefreshTokenRepository;
import io.micronaut.security.authentication.Authentication;
import io.micronaut.security.errors.OauthErrorResponseException;
import io.micronaut.security.token.event.RefreshTokenGeneratedEvent;
import io.micronaut.security.token.refresh.RefreshTokenPersistence;
import jakarta.inject.Singleton;
import org.reactivestreams.Publisher;
import reactor.core.publisher.Flux;
import reactor.core.publisher.FluxSink;

import java.time.Instant;
import java.util.Optional;

import static io.micronaut.security.errors.IssuingAnAccessTokenErrorCode.INVALID_GRANT;

@Singleton
public class CustomRefreshTokenPersistence implements RefreshTokenPersistence {


    private final RefreshTokenRepository refreshTokenRepository;
    public CustomRefreshTokenPersistence(RefreshTokenRepository refreshTokenRepository) {
        this.refreshTokenRepository = refreshTokenRepository;
    }
    @Override
    public void persistToken(RefreshTokenGeneratedEvent event) {
        if (event != null &&
                event.getRefreshToken() != null &&
                event.getAuthentication() != null &&
                event.getAuthentication().getName() != null) {
            String payload = event.getRefreshToken();

            RefreshToken refreshToken = new RefreshToken();
            refreshToken.setId(event.getAuthentication().getName());
            refreshToken.setUsername(event.getAuthentication().getName());
            refreshToken.setRevoked(false);
            refreshToken.setRefreshToken(payload);
            refreshToken.setDateCreated(Instant.now());
            refreshTokenRepository.save(refreshToken).block();
        }
    }

    @Override
    public Publisher<Authentication> getAuthentication(String refreshToken) {


        return Flux.create(emitter -> {
            Optional<RefreshToken> tokenOpt = refreshTokenRepository.findByRefreshToken(refreshToken);
            if (tokenOpt.isPresent()) {
                RefreshToken token = tokenOpt.get();
                if (token.getRevoked()) {
                    emitter.error(new OauthErrorResponseException(INVALID_GRANT, "refresh token revoked", null));
                } else {
                    emitter.next(Authentication.build(token.getUsername()));
                    emitter.complete();
                }
            } else {
                emitter.error(new OauthErrorResponseException(INVALID_GRANT, "refresh token not found", null));
            }
        }, FluxSink.OverflowStrategy.ERROR);
    }
}

File: Code Examples/JWT/Kotlin/JWTKotlinMongo/src/main/kotlin/io/hashimati/security/services/CustomRefreshTokenPersistence.kt
package io.hashimati.security.services

import io.hashimati.security.domains.RefreshToken
import io.hashimati.security.repository.RefreshTokenRepository
import io.micronaut.security.authentication.Authentication
import io.micronaut.security.errors.IssuingAnAccessTokenErrorCode
import io.micronaut.security.errors.OauthErrorResponseException
import io.micronaut.security.token.event.RefreshTokenGeneratedEvent
import io.micronaut.security.token.refresh.RefreshTokenPersistence
import jakarta.inject.Singleton
import org.reactivestreams.Publisher
import reactor.core.publisher.Flux
import reactor.core.publisher.FluxSink
import java.time.Instant

@Singleton
class CustomRefreshTokenPersistence(private val refreshTokenRepository: RefreshTokenRepository) :
    RefreshTokenPersistence {
    override fun persistToken(event: RefreshTokenGeneratedEvent) {
        if (event != null && event.refreshToken != null && event.authentication != null && event.authentication.name != null) {
            val payload = event.refreshToken
            val refreshToken = RefreshToken()
            refreshToken.id = event.authentication.name
            refreshToken.username = event.authentication.name
            refreshToken.revoked = false
            refreshToken.refreshToken = payload
            refreshToken.dateCreated = Instant.now()
            refreshTokenRepository.save(refreshToken).block()
        }
    }

    override fun getAuthentication(refreshToken: String): Publisher<Authentication> {
        return Flux.create({ emitter: FluxSink<Authentication> ->
            val tokenOpt = refreshTokenRepository.findByRefreshToken(refreshToken)
            if (tokenOpt.isPresent) {
                val (_, username, _, revoked) = tokenOpt.get()
                if (revoked!!) {
                    emitter.error(
                        OauthErrorResponseException(
                            IssuingAnAccessTokenErrorCode.INVALID_GRANT,
                            "refresh token revoked",
                            null
                        )
                    )
                } else {
                    emitter.next(Authentication.build(username))
                    emitter.complete()
                }
            } else {
                emitter.error(
                    OauthErrorResponseException(
                        IssuingAnAccessTokenErrorCode.INVALID_GRANT,
                        "refresh token not found",
                        null
                    )
                )
            }
        }, FluxSink.OverflowStrategy.ERROR)
    }
}